### PR TITLE
feat: Expose a room list filter for Invites only

### PR DESCRIPTION
**Pull request by stefanceriu**
This PR adds a new room list filter that can be used to render invites only. Meant to work together with https://github.com/matrix-org/matrix-rust-sdk/pull/3203

### CODE REVIEW

File: crates/matrix-sdk-ui/src/room_list_service/filters/invite.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    let matcher = InviteRoomMatcher {
+        is_invite: move |room| match room {
+            RoomListEntry::Filled(room_id) | RoomListEntry::Invalidated(room_id) => {
+                let room = client.get_room(room_id)?;
+                if room.state() == RoomState::Invited {
+                    Some(true)
+                } else {
+                    Some(false)
+                }
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply crates/matrix-sdk-ui/src/room_list_service/filters/invite.rs:40]: heh, of course 🤦

**Comment by stefanceriu** [comment]
> Can we have test please :-)?

hah, of course, just wanted to make sure I'm going the right path before I write any

File: crates/matrix-sdk-ui/src/room_list_service/filters/invite.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            RoomListEntry::Filled(room_id) | RoomListEntry::Invalidated(room_id) => {
+                let room = client.get_room(room_id)?;
+                Some(room.state() == RoomState::Invited)
+            }
+            _ => None,
+        },
+    };
+
+    move |room_list_entry| -> bool { matcher.matches(room_list_entry) }
+}
```
Comment (Hywan) [inline crates/matrix-sdk-ui/src/room_list_service/filters/invite.rs:43]: To be consistent with the `non_left` filter, I think we should revisit the code a little bit:

```suggestion
struct InviteRoomMatcher<F>
where
    F: Fn(&RoomListEntry) -> Option<RoomState>,
{
    state: F,
}

impl<F> InviteRoomMatcher<F>
where
    F: Fn(&RoomListEntry) -> Option<RoomState>,
{
    fn matches(&self, room: &RoomListEntry) -> bool {
        if !matches!(room, RoomListEntry::Filled(_) | RoomListEntry::Invalidated(_)) {
            return false;
        }

        if let Some(state) = (self.state)(room) {
            state == RoomState::Invited
        } else {
            false
        }
    }
}

/// Create a new filter that will accept all filled or invalidated entries, but
/// filters out rooms that are not invites (see
/// [`matrix_sdk_base::RoomState::Invited`]).
pub fn new_filter(client: &Client) -> impl Filter {
    let client = client.clone();

    let matcher = InviteRoomMatcher {
        state: move |room| {
            let room_id = room.as_room_id()?;
            let room = client.get_room(room_id)?;
            Some(room.state())
        },
    };

    move |room_list_entry| -> bool { matcher.matches(room_list_entry) }
}
```

I think consistency is preferable here, even if your code is correct!

**Review by Hywan — CHANGES_REQUESTED** [review]
The tests looks good, thanks! One minor refactoring is necessary for the sake of consistency with the `non_left` filter, but it's just a minor details :-}.

**Review by Hywan — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**